provider/server/exec.go

```diff
@@ -124,16 +123,21 @@ func (e executorExec) Exec(ctx context.Context, in io.Reader, out io.WriteCloser
 		return fmt.Errorf("machine %s has not yet been synced", machineID)
 	}
 
-	uri, err := e.Libvirt.ConnectGetUri()
+	domainXMLData, err := e.Libvirt.DomainGetXMLDesc(domain, 0)
 	if err != nil {
-		return fmt.Errorf("error getting connection uri")
+		return fmt.Errorf("failed to lookup domain: %w", err)
 	}
 
-	cmd := exec.CommandContext(ctx, e.VirshExecutable, "-c", strings.TrimSpace(uri), "console", machineID)
+	domainXML := &libvirtxml.Domain{}
+	if err := domainXML.Unmarshal(domainXMLData); err != nil {
+		return fmt.Errorf("failed to unmarshal domain: %w", err)
+	}
+
+	ttyPath := domainXML.Devices.Consoles[0].TTY
```

Comment:
i recommend check length of Consoles array, this can create panic.

Reply:
Since we set this exclusively in the domainXML during machine creation, it won't cause a panic.

Reply:
i prefer secure programming here. logic code can be change in future and new developer doesn't have to know about this code. It will throw panic which can restart app or it will cache be server. I don't know current implementation and i don't want to risk restart of whole app.

Reply:
I agree with you. Implemented a check both for nil device and empty console.